File: 2021_Kakao_Internship/Hoovely/표 편집.py
def solution(n, k, cmd):
    matrix = [True] * n
    stack = []
    for c in cmd:
        if c[0] == 'U':
            c_idx = int(c[2:])
            k -= c_idx
        elif c[0] == 'D':
            c_idx = int(c[2:])
            k += c_idx
        elif c[0] == 'C':
            if k == len(matrix)-1:  # 마지막 행을 지울때
                matrix[k] = False
                stack.append([k, matrix[k]])
                del matrix[k]
                k -= 1
            else:
                matrix[k] = False
                stack.append([k, matrix[k]])
                del matrix[k]
        elif c[0] == 'Z':
            idx, flag = stack.pop()
            matrix.insert(idx, True)

    while stack:
        idx, flag = stack.pop()
        matrix.insert(idx, False)

    answer = ''
    for i in range(n):
        if matrix[i] == True:
            answer += 'O'
        else:
            answer += 'X'

    return answer

File: 2021_Kakao_Internship/wl794/표 편집.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

string solution(int n, int k, vector<string> cmd) {
    string answer = "";

    vector<int> v(n);
    for (int i=0; i<n; i++) v[i] = i;

    stack<int> deleted;

    for (string s: cmd) {
        //answer += (k + '0');
        if (s[0] == 'U') {
            int x = stoi(s.substr(2));
            k -= x;
        } else if (s[0] == 'D') {
            int x = stoi(s.substr(2));
            k += x;
        } else if (s[0] == 'C') {
            deleted.push(v[k]);
            if (v.begin() + k != v.end())
                v.erase(v.begin() + k);
            if (k == v.size()) {
                k--;
            }
        } else {
            int del = deleted.top();
            auto pos = upper_bound(v.begin(), v.end(), del);
            v.insert(pos, del);
            deleted.pop();
            if (del < k) k++;
        }
    }

    int idx = 0;
    for (int i=0; i<n; i++) {
        if (idx == v.size()) {
            answer += ('X');
        }
        else if (v[idx] == i) {
            answer += ('O');
            idx++;
        }
        else {
            answer += ('X');
        }
    }

    return answer;
}
